libs/cassandra-util/src/Cassandra/Schema.hs

```diff
@@ -182,6 +203,49 @@ migrateSchema l o ms = do
     metaInsert :: QueryString W (Int32, Text, UTCTime) ()
     metaInsert = "insert into meta (id, version, descr, date) values (1,?,?,?)"
 
+-- | Retrieve and compare local and peer system schema versions.
+-- if they don't match, retry once per second for 30 seconds
+waitForSchemaConsistency :: Client ()
+waitForSchemaConsistency = do
+    void $ retryWhileN 30 inDisagreement getSystemVersions
+  where
+    getSystemVersions :: Client (UUID, [UUID])
+    getSystemVersions = do
+        -- These two sub-queries must be made to the same node.
+        -- (comparing local from node A and peers from node B wouldn't be correct)
+        -- using the custom 'migrationPolicy' when connecting to cassandra ensures this.
+        local <- systemLocalVersion
+        peers <- systemPeerVersions
+        case local of
+            Just localVersion -> return $ (localVersion, peers)
+            Nothing           -> liftIO $ throwIO $ ErrorCall
```

Comment:
Is there a reason why you're not simply using `error` here? If yes, I think it would be good to document it, because I spent two minutes thinking about it and haven't thought of anything.

Reply:
The only thing that comes to mind is defensive programming (e.g. what if retry doesn't work with `error`?) and I think it's a valid approach, but then the comment should say `-- Defensive programming (...)` 🙂

Btw I think retry should work with `error` but that's not exactly obvious.

Reply:
@fisx told me not to use `error` here, maybe he can elaborate his reasons? `recovering` works indeed with `error` and any `throwIO`, yes, but `retrying` (in `retryWhileN`)  doesn't catch exceptions. We're not expecting any exceptions here though - I don't mean to catch any exceptions.

Reply:
my point was that `error === unsafePerformIO . Control.Exception.throwIO . ErrorCall`, and the `unsafePerformIO` complicates the semantics of the code.  so as long as we're in `MonadIO m` anyway, it is easier to reason about what the code does if we use the latter, without the `unsafePerformIO`.  it actually changes the semantics slightly, because the error is now thrown in the execution of the action, not in the construction.  but this is closer to at least my own intuition when i read this function.

does this make sense?  perhaps we should have

```haskell
errorIO :: (Show msg, MonadIO m) => msg -> m a
errorIO = liftIO . Control.Exception.throwIO . ErrorCall
```

somewhere in the libraries?